pkg/apis/apis/v1alpha1/types_apiresourceschema.go

```diff
@@ -163,3 +163,88 @@ func (v *APIResourceVersion) SetSchema(schema *apiextensionsv1.JSONSchemaProps)
 	v.Schema.Raw = raw
 	return nil
 }
+
+// +crd
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster,categories=kcp
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+
+// APIConversion contains rules to convert between different API versions in an APIResourceSchema. The name must match
+// the name of the APIResourceSchema for the conversions to take effect.
+type APIConversion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec holds the desired state.
+	Spec APIConversionSpec `json:"spec"`
+}
+
+// APIConversionSpec contains rules to convert between different API versions in an APIResourceSchema.
+type APIConversionSpec struct {
+	// +required
+	// +listType=map
+	// +listMapKey=from
+	// +listMapKey=to
+
+	// conversions specify rules to convert between different API versions in an APIResourceSchema.
+	Conversions []APIVersionConversion `json:"conversions"`
+}
+
+// APIVersionConversion contains rules to convert between two specific API versions in an
+// APIResourceSchema. Additionally, to avoid data loss when round-tripping from a version that
+// contains a new field to one that doesn't and back again, you can specify a list of fields to
+// preserve (these are stored in annotations).
+type APIVersionConversion struct {
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// from is the source version.
+	From string `json:"from"`
+
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern=^v[1-9][0-9]*([a-z]+[1-9][0-9]*)?$
+
+	// to is the target version.
+	To string `json:"to"`
+
+	// +required
+	// +listType=map
+	// +listMapKey=to
+
+	// rules contains field-specific conversion expressions.
+	Rules []APIConversionRule `json:"rules"`
+
+	// +optional
+
+	// preserve contains a list of paths to fields to preserve, relative to the root of the object, such as
+	// 'spec.name.first'.
+	Preserve []string `json:"preserve,omitempty"`
+}
+
+// APIConversionRule specifies how to convert a single field.
+type APIConversionRule struct {
+	// +required
+
+	// rule is a CEL expression specifying the data to use from the source version to populate into the field referenced
+	// by 'to' in the target version. References to fields are relative to the 'from' version, such as 'v1.spec.name'.
+	Rule string `json:"rule"`
```

Comment:
am not convinced of the factoring. Would rather prefer this:

```yaml
from: v1
to: v2
rule:
   field: <extended-json-path>
   destination: <extended-json-path> # defaults to field
   transformation: <CEL> # default to self
```

This naturally extends to arrays and maps:
```yaml
from: v1
to: v2
rule:
   field: .spec.versions[*].labels
   destination: .spec.somewhere.else.versions[*].versionLabels
   transformation: self
```
Here the sequence of quantifiers `[*]` must match.